File: apex_dma/Math.cpp
#include "Math.h"

void Math::NormalizeAngles(QAngle& angle)
{
	while (angle.x > 89.0f)
		angle.x -= 180.f;

	while (angle.x < -89.0f)
		angle.x += 180.f;

	while (angle.y > 180.f)
		angle.y -= 360.f;

	while (angle.y < -180.f)
		angle.y += 360.f;
}

QAngle Math::CalcAngle(const Vector& src, const Vector& dst)
{
	QAngle angle = QAngle();
	SVector delta = SVector((src.x - dst.x), (src.y - dst.y), (src.z - dst.z));

	double hyp = sqrt(delta.x*delta.x + delta.y * delta.y);

	angle.x = atan(delta.z / hyp) * (180.0f / M_PI);
	angle.y = atan(delta.y / delta.x) * (180.0f / M_PI);
	angle.z = 0;
	if (delta.x >= 0.0) angle.y += 180.0f;

	return angle;
}

double Math::GetFov(const QAngle& viewAngle, const QAngle& aimAngle)
{
	QAngle delta = aimAngle - viewAngle;
	NormalizeAngles(delta);

	return sqrt(pow(delta.x, 2.0f) + pow(delta.y, 2.0f));
}

double Math::DotProduct(const Vector& v1, const float* v2)
{
	return v1.x * v2[0] + v1.y * v2[1] + v1.z * v2[2];
}

float Math::SmoothStep(float edge0, float edge1, float x)
{
	// Scale, bias and saturate x to 0..1 range
	x = (x - edge0) / (edge1 - edge0);
	x = x < 0 ? 0 : (x > 1 ? 1 : x);
	// Evaluate polynomial
	return x * x * (3 - 2 * x);
}

bool Math::IsLineIntersectingSphere(const Vector& start, const Vector& end, const Vector& sphereCenter, float radius)
{
	const float CHECK_RADIUS = radius + 10.0f;
	const float INTERSECT_RADIUS = radius;

	auto getDistanceSqr = [](const Vector& point, const Vector& center) {
		return (point.x - center.x) * (point.x - center.x)
			+ (point.y - center.y) * (point.y - center.y)
			+ (point.z - center.z) * (point.z - center.z);
	};

	float startDistSqr = getDistanceSqr(start, sphereCenter);
	float endDistSqr = getDistanceSqr(end, sphereCenter);
	float checkRadiusSqr = CHECK_RADIUS * CHECK_RADIUS;

	// If one point is inside and the other is outside, the line must intersect the sphere surface
	if ((startDistSqr <= checkRadiusSqr) ^ (endDistSqr <= checkRadiusSqr)) {
		return true;
	}

	// If both points are inside, they can see each other through the dome
	if (startDistSqr <= checkRadiusSqr && endDistSqr <= checkRadiusSqr) {
		return false;
	}

	// Ray-Sphere intersection for both points outside
	Vector dir = {
		end.x - start.x,
		end.y - start.y,
		end.z - start.z
	};

	Vector oc = {
		start.x - sphereCenter.x,
		start.y - sphereCenter.y,
		start.z - sphereCenter.z
	};

	float a = dir.x * dir.x + dir.y * dir.y + dir.z * dir.z;
	float b = 2.0f * (oc.x * dir.x + oc.y * dir.y + oc.z * dir.z);
	float c = (oc.x * oc.x + oc.y * oc.y + oc.z * oc.z) - (INTERSECT_RADIUS * INTERSECT_RADIUS);

	float discriminant = b * b - 4 * a * c;

	if (discriminant < 0) {
		return false;
	}

	float t1 = (-b - sqrt(discriminant)) / (2 * a);
	float t2 = (-b + sqrt(discriminant)) / (2 * a);

	return (t1 >= 0 && t1 <= 1) || (t2 >= 0 && t2 <= 1);
}

File: apex-rust/host/src/math.rs
use std::ops::Sub;

#[derive(Copy, Clone, Debug, Default, PartialEq)]
pub struct Vector3 {
    pub x: f32,
    pub y: f32,
    pub z: f32,
}

impl Vector3 {
    pub fn new(x: f32, y: f32, z: f32) -> Self {
        Self { x, y, z }
    }

    pub fn dist_to(&self, other: &Self) -> f32 {
        ((self.x - other.x).powi(2) + (self.y - other.y).powi(2) + (self.z - other.z).powi(2)).sqrt()
    }

    pub fn length(&self) -> f32 {
        (self.x.powi(2) + self.y.powi(2) + self.z.powi(2)).sqrt()
    }
}

impl Sub for Vector3 {
    type Output = Self;
    fn sub(self, other: Self) -> Self {
        Self::new(self.x - other.x, self.y - other.y, self.z - other.z)
    }
}

pub fn normalize_angles(mut angle: Vector3) -> Vector3 {
    while angle.x > 89.0 {
        angle.x -= 180.0;
    }
    while angle.x < -89.0 {
        angle.x += 180.0;
    }
    while angle.y > 180.0 {
        angle.y -= 360.0;
    }
    while angle.y < -180.0 {
        angle.y += 360.0;
    }
    angle
}

pub fn calc_angle(src: Vector3, dst: Vector3) -> Vector3 {
    let delta = src - dst;
    let hyp = (delta.x.powi(2) + delta.y.powi(2)).sqrt();

    let mut angle = Vector3::new(
        (delta.z / hyp).atan().to_degrees(),
        (delta.y / delta.x).atan().to_degrees(),
        0.0
    );

    if delta.x >= 0.0 {
        angle.y += 180.0;
    }

    normalize_angles(angle)
}

pub fn get_fov(view_angle: Vector3, aim_angle: Vector3) -> f32 {
    let delta = normalize_angles(aim_angle - view_angle);
    (delta.x.powi(2) + delta.y.powi(2)).sqrt()
}
